src/python/pants/task/fmt_task_mixin.py

```diff
@@ -0,0 +1,19 @@
+# coding=utf-8
+# Copyright 2018 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import (absolute_import, division, generators, nested_scopes, print_function,
+                        unicode_literals, with_statement)
+
+from pants.task.goal_options_mixin import GoalOptionsMixin
+from pants.task.target_restriction_mixins import (HasSkipAndTransitiveOptionsMixin,
+                                                  SkipAndTransitiveOptionsRegistrarBase)
+
+
+class FmtGoalOptionsRegistrar(SkipAndTransitiveOptionsRegistrarBase):
+  options_scope = 'fmt'
+
+
+class FmtTaskMixin(GoalOptionsMixin, HasSkipAndTransitiveOptionsMixin):
+  """A mixin to combine with code formatting tasks."""
+  goal_options_registrar_cls = FmtGoalOptionsRegistrar
```

Comment:
The idea of a task installing a class indicating which goal it is in feels like a departure from our current "tasks may be installed in multiple goals by register.py" model. Is that intentional?

Alternatively, how could this be better aligned with Subsystems? The fact that Subsystems are explicitly referenced, and not merged into the options namespace of the task is pretty nice... I'm worried that the mixins end up looking a bit like a forest.

One relatively minor edit might be to have the mixins override `def subsystem_dependencies` to insert a Subsystem dependency, and then depend on that Subsystem for the option. A bit less magic perhaps?

Reply:
Yeah, it's not ideal. Thing is, that abstraction will be a bit broken by *any* mechanism that allows goal-level options, because something has to register them once at the goal level, and then each task has to know to *use* the option. So in some sense the task knows its goal, or at least it knows it's in some goal that has those options registered.

Using a subsystem would require changing the names of all those task-level options, which not only imposes a massive deprecation cycle, but also makes the names clunkier. (I don't think we can register a subsystem with the same scope as a goal, that is asking for all kinds of trouble). I kinda like that this is lightweight, and just leverages our existing ability to register options recursively. 

So how about this - instead of the mixin specifying the goal it's in, it specifies the options it has access to. I.e., we get rid of `goal_options_registrar_cls`? That's really only needed for test setup anyway, and we can work around that some other way. So now there's no link between the task and its goal name, so you can register the task in multiple goals (as long as the goal provides the options, of course)?